### PR TITLE
feat: add `lapack/base/dge-trans`

### PR DESCRIPTION
**Pull request by Pranavchiku**
Towards #2464.

## Description

> What is the purpose of this pull request?

This pull request exposes utility function `dge_trans` as `lapack/base/dge-trans`.

## Related Issues

> Does this pull request have any related issues?

NA

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/lapack/base/transpose/README.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# transpose
+
+> Convert input general matrix from row-major order to column-major layout or vice versa.
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/transpose/README.md]:
```suggestion
> Convert input general matrix from row-major to column-major layout or vice versa.
```

File: lib/node_modules/@stdlib/lapack/base/transpose/package.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    "windows"
+  ],
+  "keywords": [
+    "stdlib",
+    "stdmath",
+    "mathematics",
+    "math",
+    "lapack",
+    "transpose",
+    "copy",
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/transpose/package.json]:
```suggestion
```

**Comment by kgryte** [comment]
@Pranavchiku Is there a reason to name the package as `transpose` rather than `dge-trans`? We'll need separate packages for single-precision and complex dtypes.

**Comment by Pranavchiku** [comment]
I see, I initially used snake case and it threw me lint error. Also, I am unable to understand sense of `dge-*`, can we not have `dtrans`?

**Comment by Pranavchiku** [comment]
Is it double precision general matrix transpose? If yes, then it makes sense.

File: lib/node_modules/@stdlib/lapack/base/dge-trans/README.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+# dge-trans
+
+> Convert input general matrix from row-major to column-major layout or vice versa.
+
+<section class = "usage">
+
+## Usage
+
+```javascript
+var transpose = require( '@stdlib/lapack/base/dge-trans' );
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]:
```suggestion
var dge-trans = require( '@stdlib/lapack/base/dge-trans' );
```

Shall I do the following or `transpose` is fine?

**Comment by kgryte** [comment]
> Is it double precision general matrix transpose? If yes, then it makes sense.

Yes, I believe so. Not sure why they didn't do `dgetrans`, but 🤷‍♂️ .

File: lib/node_modules/@stdlib/lapack/base/dge-trans/README.md (hunk 3 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]: You cannot do `dge-trans`. Do `dgetrans`?

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]: Done :)

Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]: @Pranavchiku No. The package name was fine and as it should be `dge-trans`. The variable name cannot have a hyphen.

Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]: I mean, we are also free to name things as we want. But my comment was specifically directed to `var dge-trans`, which is not valid JS.

Comment (Pranavchiku) [reply lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]: So shall I revert back to package `dge-trans`?

Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dge-trans/README.md]: If you don't mind. 😬

File: lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	var sa;
+	var so;
+	var i;
+	var j;
+	var x;
+	var y;
+
+	if ( isRowMajor( [ strideA1, strideA2 ] ) ) {
+		x = M;
+		y = N;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js]: Use `m` and `n`. Using `x` and `y` given that they commonly refer to arrays is confusing here.

File: lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	} else {
+		x = N;
+		y = M;
+	}
+
+	for ( i = 0; i < y; i++ ) {
+		sa = offsetA + ( i * strideA2 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js]: Don't use `s*` for "pointers". We use `s*` to refer to strides in other implementations. The common convention is `ia` and `io` when applied here.

File: lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+*/
+function dgetrans( M, N, A, strideA1, strideA2, offsetA, out, strideO1, strideO2, offsetO ) {
+	var sa;
+	var so;
+	var i;
+	var j;
+	var x;
+	var y;
+
+	if ( isRowMajor( [ strideA1, strideA2 ] ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js]: It looks like you are attempting to perform loop interchange here, but I believe you also need to swap strides.

File: lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js (hunk 5 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/lapack/base/dgetrans/lib/base.js]: You can also use `oa` and `oo`.

**Review by kgryte — CHANGES_REQUESTED** [review]
Left some initial comments.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks for working on this, @Pranavchiku.

Note that the tests were rather spartan. I would argue it is better to create tests which build on one another, rather than creating a single test which attempts to do all the things. By isolating one behavioral aspect at a time, we can better ensure that an implementation behaves as expected. For combining all the things into one, it can be harder to disentangle which aspect is causing a test to fail.

**Comment by kgryte** [comment]
In principle, we could implement matrix transposition using loop tiling. Loop tiling would only benefit when layout access patterns for `A` and `out` differ. E.g., if both `A` and `out` are the same (e.g., row-major), because we are taking the transpose, the access pattern for `A` will be column-major. In this instance, loop tiling can be beneficial. Otherwise, a naive nested loop is cache optimal. Regardless, I'll leave implementing loop tiling to a follow-up PR.

**Pull request merged**